python/cuml/dask/common/input_utils.py

```diff
@@ -0,0 +1,314 @@
+        self.ranks = None
+        self.parts_to_sizes = None
+        self.total_rows = None
+
+    """ Class methods for initalization """
+
+    @classmethod
+    def single(cls, data, client=None):
+        gpu_futures = client.sync(_extract_partitions, data, client)
+
+        worker_to_parts = _workers_to_parts(gpu_futures)
+
+        workers = list(map(lambda x: x[0], worker_to_parts.items()))
+
+        datatype = 'cudf' if isinstance(data, dcDataFrame) else 'cupy'
+
+        return MGData(gpu_futures=gpu_futures, worker_to_parts=worker_to_parts,
+                      workers=workers, datatype=datatype, multiple=False,
+                      client=client)
+
+    @classmethod
+    def colocated(cls, data, force=False, client=None):
+        gpu_futures = client.sync(_extract_colocated_partitions,
+                                  data[0], data[1], client)
+
+        workers = list(gpu_futures.keys())
+
+        datatype = 'cudf' if isinstance(data[0], dcDataFrame) else 'cupy'
+
+        return MGData(gpu_futures=gpu_futures, workers=workers,
+                      datatype=datatype, multiple=True, client=client)
+
+    """ Methods to calculate further attributes """
+
+    def calculate_worker_and_rank_info(self, comms):
+        self.worker_info = comms.worker_info(comms.worker_addresses)
+        self.ranks = dict()
+        for w, futures in self.gpu_futures.items():
+                self.ranks[w] = self.worker_info[w]["r"]
+
+    def calculate_parts_to_sizes(self, comms=None, ranks=None):
+        if self.worker_info is None and comms is not None:
+            self.calculate_worker_and_rank_info(comms)
+
+        ranks = self.ranks if ranks is None else ranks
+
+        self.total_rows = 0
+
+        self.parts_to_sizes = dict()
+
+        # func = _get_rows_from_colocated_tuple if self.multiple is True else \
+        #     _get_rows_from_single_obj
+
+        key = uuid1()
+
+        if self.multiple:
+            for w, futures in self.gpu_futures.items():
+                parts = [(self.client.submit(
+                    _get_rows_from_colocated_tuple,
+                    future,
+                    workers=[w],
+                    key="%s-%s" % (key, idx)).result())
+                    for idx, future in enumerate(futures)]
+
+                self.parts_to_sizes[self.worker_info[w]["r"]] = parts
+                for p in parts:
+                    self.total_rows = self.total_rows + p
+
+        else:
+                self.parts_to_sizes = [(ranks[wf[0]], self.client.submit(
+                                       _get_rows_from_single_obj,
+                                       wf[1],
+                                       workers=[wf[0]],
+                                       key="%s-%s" % (key, idx)).result())
+                                       for idx,
+                                       wf in enumerate(self.gpu_futures)]
+
+                self.total_rows = reduce(lambda a, b:
```

Comment:
This is one of those things where I'm really pro-list-comprehension:
 `sum([x[1] for x in self.parts_to_sizes])`